### PR TITLE
feature/GSYE-386: Add accumulators for forward markets

### PR DESCRIPTION
**Pull request by mrsaemir**
As requested by [GSYE-386](https://gridsingularity.atlassian.net/browse/GSYE-386), we need accumulator functions to calculate global energy trade profiles for each device in forward markets.

## Proposed changes:
- [X] Add EBEnergyTradeProfileAccumulator class
- [X] Enhance performance of MarketResultsAggregator._check_timeslots
- [x] Add unit tests

## Reproduction steps:
```python
if __name__ == '__main__':

    import json
    import pprint
    from pendulum import duration, DateTime
    from gsy_framework.sim_results.aggregate_results import MarketResultsAggregator
    from gsy_framework.sim_results.electric_blue.accumulators import EBEnergyTradeProfileAccumulator

    with open("results", "r") as file:
        slot_data = file.read().split("<--->")
        data = [json.loads(d) for d in slot_data if d]

    mrh = MarketResultsAggregator(
        resolution=duration(hours=6),
        simulation_slot_length=duration(minutes=15),
        accumulators={"energy_trade_profile": EBEnergyTradeProfileAccumulator()}
    )

    for slot_data in data:
        if not slot_data["current_market"]:
            continue
        c_timeslot = DateTime.fromisoformat(slot_data["current_market"])
        for uuid, market in slot_data["simulation_raw_data"].items():
            day_forward_forward_market_stats = market.get("forward_market_stats",
                                                          {}).get("4", {})
            mrh.update(c_timeslot, day_forward_forward_market_stats)

            for i in mrh.generate():
                pprint.pprint(i)
```

## Example results:
```
{'accumulated_results': {'energy_trade_profile': {'6cba7aaa-e30d-4c94-977a-4a1fb2ceb87a': {'total_buy_trade_count': 0,
                                                                                           'total_earned_eur': 665280.0,
                                                                                           'total_energy_bought': 0,
                                                                                           'total_energy_consumed': 0,
                                                                                           'total_energy_produced': 0,
                                                                                           'total_energy_sold': 22176,
                                                                                           'total_sell_trade_count': 2016,
                                                                                           'total_spent_eur': 0},
                                                  'e5e80a9e-288c-4114-be4e-dd330b673016': {'total_buy_trade_count': 2016,
                                                                                           'total_earned_eur': 0,
                                                                                           'total_energy_bought': 22176,
                                                                                           'total_energy_consumed': 0,
                                                                                           'total_energy_produced': 0,
                                                                                           'total_energy_sold': 0,
                                                                                           'total_sell_trade_count': 0,
                                                                                           'total_spent_eur': 665280.0}}},
 'aggregated_results': {},
 'start_time': DateTime(2022, 8, 22, 0, 0, 0)}
{'accumulated_results': {'energy_trade_profile': {'6cba7aaa-e30d-4c94-977a-4a1fb2ceb87a': {'total_buy_trade_count': 0,
                                                                                           'total_earned_eur': 2980800.0,
                                                                                           'total_energy_bought': 0,
                                                                                           'total_energy_consumed': 0,
                                                                                           'total_energy_produced': 0,
                                                                                           'total_energy_sold': 99360,
                                                                                           'total_sell_trade_count': 4320,
                                                                                           'total_spent_eur': 0},
                                                  'e5e80a9e-288c-4114-be4e-dd330b673016': {'total_buy_trade_count': 4320,
                                                                                           'total_earned_eur': 0,
                                                                                           'total_energy_bought': 99360,
                                                                                           'total_energy_consumed': 0,
                                                                                           'total_energy_produced': 0,
                                                                                           'total_energy_sold': 0,
                                                                                           'total_sell_trade_count': 0,
                                                                                           'total_spent_eur': 2980800.0}}},
 'aggregated_results': {},
 'start_time': DateTime(2022, 8, 22, 6, 0, 0)}
 ```

[results.zip](https://github.com/gridsingularity/gsy-framework/files/9430549/results.zip)

### CODE REVIEW

**Review by spyrostz — APPROVED** [review]
LGTM, after the addition of 1 TODO: in the future this needs to also aggregate the installed energy capacity. No need to do anything for this since it depends on Pierpaolo's PR, however would be good to have as TODO here in order to not forget

**Review by fievelk — APPROVED** [review]
I'm a bit confused by the fact that these components are not used in the code (or maybe I just couldn't find the place where they are used). I left some comments about tests, but I'll give my LGTM.

File: gsy_framework/sim_results/aggregate_results.py [hunk 1]
```diff
@@ -53,11 +55,11 @@
 RESOLUTION_AGGREGATIONS = {
     AggregationResolution.RES_1_MONTH: {
         "aggregators": {},
-        "accumulators": {}
+        "accumulators": {"energy_trade_profile": EBEnergyTradeProfileAccumulator()}
```
[the remark this answers is not in the document]
Comment (mrsaemir) [reply gsy_framework/sim_results/aggregate_results.py:58]: Well, yes. This is supposed to be used by Kamil in a future PR.

File: gsy_framework/sim_results/electric_blue/accumulators.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            current_results[seller_id]["total_energy_sold"] += trade["energy"]
+            current_results[seller_id]["total_earned_eur"] += trade["price"]
+            current_results[seller_id]["total_sell_trade_count"] += 1
+
+            # BUYER
+            current_results[buyer_id]["total_energy_bought"] += trade["energy"]
+            current_results[buyer_id]["total_spent_eur"] += trade["price"]
+            current_results[buyer_id]["total_buy_trade_count"] += 1
+
+        return current_results
```
[the remark this answers is not in the document]
Comment (mrsaemir) [reply gsy_framework/sim_results/electric_blue/accumulators.py:56]: In the beginning I wanted to implement both the aggregator and the accumulator in a single PR, but then I decided to do it in a separate one to make things simpler. The empty module is filled using my other PR.
Thanks

File: gsy_framework/sim_results/aggregate_results.py [hunk 3]
```diff
@@ -23,6 +23,8 @@
 from gsy_framework.enums import AggregationResolution, AvailableMarketTypes
 from gsy_framework.sim_results.area_throughput_stats import AreaThroughputStats
 from gsy_framework.sim_results.device_statistics import DeviceStatistics
+from gsy_framework.sim_results.electric_blue.accumulators import \
+    EBEnergyTradeProfileAccumulator
```
[the remark this answers is not in the document]
Comment (mrsaemir) [reply gsy_framework/sim_results/aggregate_results.py:27]: Oops. I missed that. I'll fix it in the next PR very soon.

File: tests/test_sim_results/electric_blue/test_accumulators.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+from pendulum import DateTime, duration
+
+from gsy_framework.sim_results.aggregate_results import MarketResultsAggregator
+from gsy_framework.sim_results.electric_blue.accumulators import \
+    EBEnergyTradeProfileAccumulator
+from tests.test_sim_results.test_aggregate_results import (
+    MARKET_TIMESLOTS, ORDERS_PER_MARKET_SLOT, gen_market_stats)
```
[the remark this answers is not in the document]
Comment (mrsaemir) [reply tests/test_sim_results/electric_blue/test_accumulators.py:7]: There's no modification to them. I just needed their value.

File: tests/test_sim_results/electric_blue/test_accumulators.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+
+class TestEBAccumulators:
```
[the remark this answers is not in the document]
Comment (mrsaemir) [reply tests/test_sim_results/electric_blue/test_accumulators.py:10]: The `MarketResultsAggregator` provided me some data preparation steps which I had to do manually. But I agree with you that the Aggr/Accu should be tested in isolation.

**Pull request merged**